src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs

```diff
@@ -148,11 +149,16 @@ public void Apply(ApplicationModel application)
                     throw new ArgumentNullException(nameof(application));
                 }
 
-                foreach (var controller in application.Controllers)
+                // Create copies of collections of controllers, actions and parameters as users could modify
+                // these collections from within the convention itself.
+                var controllers = application.Controllers.ToArray();
```

Comment:
Guessing you also looked at Razor Pages conventions to make sure it doesn't need something like this

Reply:
Yeah I think pages also needs this too, it has parameters at least.

Reply:
Checked with @pranavkm regarding this. We do not have similar concerns with conventions in RazorPages. https://github.com/aspnet/Mvc/issues/6935 issue is for handling parameters and has not been implemented yet.